--- search-items.c.orig	2009-03-08 03:05:31.000000000 +0100
+++ search-items.c	2009-03-08 03:06:02.000000000 +0100
@@ -434,7 +434,7 @@
             if(passed)
                 printf(", ");
             
-            printf(ret);
+            printf("%s", ret);
             passed = 1;
         }
 	test++;
--- ma-search-users.c.orig	2009-03-08 03:07:32.000000000 +0100
+++ ma-search-users.c	2009-03-08 03:08:25.000000000 +0100
@@ -39,7 +39,7 @@
             if(mult != 0)
                 printf(", %s", cwd);
             else {
-                printf(cwd);
+                printf("%s", cwd);
                 mult = 1;
             }
         }
@@ -121,7 +121,7 @@
             if(mult != 0)
                 printf(", %s", cwd);
             else {
-                printf(cwd);
+                printf("%s", cwd);
                 mult = 1;
             }
         }
